xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/VaapiEGL.cpp

```diff
@@ -445,8 +468,279 @@ bool CVaapiTexture::TestInteropHevc(VADisplay vaDpy, EGLDisplay eglDisplay)
+{
+  GLuint texture;
+  glGenTextures(1, &texture);
+  glBindTexture(m_interop.textureTarget, texture);
+  m_interop.glEGLImageTargetTexture2DOES(m_interop.textureTarget, image);
+  glBindTexture(m_interop.textureTarget, 0);
+  return texture;
+}
+
+bool CVaapi2Texture::Map(CVaapiRenderPicture *pic)
+{
+  if (m_vaapiPic)
+    return true;
+
+  m_vaapiPic = pic;
+  m_vaapiPic->Acquire();
+
+  VAStatus status;
+
+  VADRMPRIMESurfaceDescriptor surface;
+
+  status = vaExportSurfaceHandle(pic->vadsp, pic->procPic.videoSurface,
+                                 VA_SURFACE_ATTRIB_MEM_TYPE_DRM_PRIME_2,
+                                 VA_EXPORT_SURFACE_READ_ONLY | VA_EXPORT_SURFACE_SEPARATE_LAYERS,
+                                 &surface);
+
+  if (status != VA_STATUS_SUCCESS)
+  {
+    CLog::LogFunction(LOGWARNING, "CVaapi2Texture::Map", "vaExportSurfaceHandle failed - Error: %s (%d)", vaErrorStr(status), status);
+    return false;
+  }
+
+  // Remember fds to close them later
+  if (surface.num_objects > m_drmFDs.size())
+    throw std::logic_error("Too many fds returned by vaExportSurfaceHandle");
+
+  for (int object = 0; object < surface.num_objects; object++)
+  {
+    m_drmFDs[object].attach(surface.objects[object].fd);
+  }
+
+  status = vaSyncSurface(pic->vadsp, pic->procPic.videoSurface);
+  if (status != VA_STATUS_SUCCESS)
+  {
+    CLog::LogFunction(LOGERROR, "CVaapiTexture::MapEsh", "vaSyncSurface - Error: %s (%d)", vaErrorStr(status), status);
+    return false;
+  }
+
+  m_textureSize.Set(surface.width, surface.height);
+
+  for (int layerNo = 0; layerNo < surface.num_layers; layerNo++)
+  {
+    int plane = 0;
+    auto const& layer = surface.layers[layerNo];
+    if (layer.num_planes != 1)
+    {
+      CLog::LogFunction(LOGDEBUG, "CVaapiTexture::MapEsh", "DRM-exported layer has %d planes - only 1 supported", layer.num_planes);
+      return false;
+    }
+    auto const& object = surface.objects[layer.object_index[plane]];
+
+    MappedTexture* texture{};
+    EGLint width{m_textureSize.Width()};
+    EGLint height{m_textureSize.Height()};
+
+    switch (surface.num_layers)
+    {
+      case 2:
+        switch (layerNo)
+        {
+          case 0:
+            texture = &m_y;
+            break;
+          case 1:
+            texture = &m_vu;
+            if (surface.fourcc == VA_FOURCC_NV12 || surface.fourcc == VA_FOURCC_P010 || surface.fourcc == VA_FOURCC_P016)
+            {
+              // Adjust w/h for 4:2:0 subsampling on UV plane
+              width = (width + 1) >> 1;
+              height = (height + 1) >> 1;
+            }
+            break;
+          default:
+            throw std::logic_error("Impossible layer number");
+        }
+        break;
+      default:
+        CLog::LogFunction(LOGDEBUG, "CVaapiTexture::MapEsh", "DRM-exported surface %d layers - only 2 supported", surface.num_layers);
+        return false;
+    }
+
+    EGLint attribs[13 + 4 /* modifiers */] = {
```

Comment:
drop those ugly comments interweaved with code. if required add at end of line

Reply:
done